text/1162449/undirected_ring_stages_with_feedback.tex

```diff
@@ -0,0 +1,99 @@
+Then regardless if any message was sent from $n(i)$, it will change its state to \textit{A-candidate}.\\
+If $i=vl=vr$ then $n(i)$ is the only \textit{E-candidate} in the ring. It should become the \textit{Leader} and send \textit{T-msg} in one of its directions.\\
+\\
+Each \textit{A-candidate} awaits approval from both of its active neighbours, meaning, it waits for message $(\textit{A-msg},i)$ from both sides. Afterwards, it changes its status to \textit{E-candidate} to participate in the next round.\\
+If \textit{A-candidate} receives $(\textit{E-msg},i)$, it should treat it as a negative feedback, it should become \textit{passive} and pass the received message.\\
+\\
+\textit{passive} node passes every message, except for potential $(\textit{A-msg},i)$ which is no longer needed.\\
+If it receives $(\textit{T-msg},val)$, it learns that $n(val)$ is the leader, passes the message and finishes.\\
+\\
+\textit{Leader} finishes when it receives $(\textit{T-msg},i)$.\\
+\\
+It shall be noted that receiving \textit{T-msg} can be used to obtain common sense of direction in the ring.
+\newpage
+\subsection*{Correctness}
+Let $j=max\{i | n(i) \text{ participates in the algorithm}\}$. \\
+Then from the voting and checking procedure follows that at each electoral round $n(j)$ will defeat both of its immediate active neighbours; hence the number of \textit{passive} nodes increases monotonically. At the same time $n(j)$ will never be defeated (receive negative feedback). This implies that at some round $n(j)$ will be the only active node and both of its \textit{E-msg} messages will return to it, making it the leader.\\
+Having been elected $n(j)$ can send a termination order \textit{T-msg} around the ring along with its identifier to notify other nodes of its election.
+\subsection*{Complexity analysis}
+By $m_i$ let's denote set of nodes participating in $i$-th election (number of \textit{E-candidates}).\\
+Let $r_j(n(i))$ where $n(i)\in m_j$ be the immediate neighbour to the left of $n(i)$ who is still active in $j$-th round, symmetrically we define $l_j(n(i)$.
+\begin{lemma}[1]
+    $|m_j|\geq 3|m_{j+1}|$ $\forall j$
+\end{lemma}
+\begin{proof}
+If a node $n(i)\in m_j$ survives the $j$-th round, then it means that it must have received positive acknowledgment from both $r_j(n(i)$ and $l_j(n(i)$.
+Therefore, because of the voting procedure, negative feedback will be received by $l^{2}_{j}(n(i)$, $l_{j}(n(i)$, $r_{j}(n(i)$ and $r^{2}_{j}(n(i)$. Thus, for each surviving node, at least 2 were eliminated. 
+Hence, at most a third of the processors from $m_j$ survive the $j$-th round.
+\end{proof}
+\begin{lemma}[2]
+    number of electoral rounds $\leq \lceil\log_{3}n\rceil$
+\end{lemma}
+\begin{proof}
+The proof follows from Lemma (1).
+\end{proof}
+\paragraph{}
+We can now notice that each round every 2 immediate active nodes will exchange at most 3 messages. That is, 2 \textit{E-msg} messages and no more than one \textit{A-msg}.\\
```

Comment:
Each number should be put in the math mode to be formatted (e.g. spaced) properly e.g. $2$.